Port faerie to the targeting crate.

sunfishcode opened the pull request:
`targeting` is a new crate that provides a vocabulary for describing targets,
including their architecture and binary file format. The patch here replaces
faerie's `Target` enum and `Object` trait with targeting's `Triple` struct.

This has the following advantages:
 - Artifact::emit and Artifact::write no longer have a type parameter, so
   users of faeiry don't have to add extra abstraction in order to support
   multiple formats dynamically (as Cretonne currently does [here]).

 - targeting is rich enough that compilers can use it directly. It supports
   the same set of target names that rustc uses. This eliminates the need to
   translate between different target enums when using faerie.

[here]: https://github.com/cretonne/cretonne/blob/master/lib/faerie/src/container.rs#L7

One downside is that this ends up requring the binary format to be specified
in the `Artifact` constructor, rather than being a parameter to `emit` or
`write` at the end. However, if this turns out to be important in practice,
it wouldn't be difficult to add a way to mutate an `Artifact`'s target.

This PR currently uses a git url dependency. I'm planning to publish a crates.io
package, but I'm hoping to get some feedback, including potentially on the name of
the package :-). I'm not settled on `targeting` yet.

m4b commented on tests/artifact.rs:
```diff
@@ -1,11 +1,14 @@
 fn duplicate_declarations_are_ok() {
 
-    let mut obj = Artifact::new(Target::X86_64, "t.o".into());
+    let mut obj = Artifact::new(triple!("x86_64"), "t.o".into());
```
Where is the rest of the triple here ? Does it have a default ?

m4b commented on README.md:
```diff
@@ -74,8 +74,8 @@ obj.link(Link { from: "main", to: "printf", at: 29 })?;
 obj.link(Link { from: "deadbeef", to: "DEADBEEF", at: 7 })?;
 
-// Finally, we write which object file we desire
-obj.write::<Elf>(file)?;
```
Not going to lie, this was one of my favorite features and sad it might go away.

I’m not sure how often the use case was where the flexibility was required to be lazy about final binary target, but I nevertheless thought it was awesome and unique.

m4b commented on src/bin/main.rs:
```diff
@@ -51,7 +53,18 @@ pub struct Args {
 fn run (args: Args) -> Result<(), Error> {
     let file = File::create(Path::new(&args.filename))?;
-    let mut obj = ArtifactBuilder::new(Target::X86_64)
+    let target = Triple {
```
This is awesome; no matter the decision something like this is needed eventually and I think it’s appropriate for it to be a separate crate

m4b commented on src/bin/main.rs:
```diff
@@ -51,7 +53,18 @@ pub struct Args {
+        vendor: Vendor::Unknown,
+        operating_system: OperatingSystem::Unknown,
+        environment: Environment::Unknown,
+        binary_format: if args.mach {
```
That this is a field also gives me hope that the binary format isn’t conflated/tightly coupled with the with the OS target, etc.

In particular, this is one aspect of traditional string triples that really bothers me and I hope faerie can (still) be more flexible about - eg I tell you the machine target and the binary format and don’t need to specify anything else.

m4b commented on src/elf.rs:
```diff
@@ -687,24 +714,22 @@ impl<'a> Elf<'a> {
-        let mut buffer = Cursor::new(Vec::new());
-        elf.write(&mut buffer)?;
-        Ok(buffer.into_inner())
+pub fn to_bytes(artifact: &Artifact) -> Result<Vec<u8>, Error> {
```
Shouldn’t this be a method in Elf?

Similarly could leave object trait and have all the targets implement this method, along with emit ? Or perhaps not necessary, dunno

m4b commented on src/target.rs:
```diff
@@ -2,26 +2,19 @@
-        }
-    }
+pub fn make_ctx(target: &Triple) -> Ctx {
+    let container_size = match target.pointer_width() {
```
This is cool, I like this, good to start building this in for the future

pchickey commented:
I think `target_lexicon` is a better name than `targeting` for the reasons you cited

sunfishcode commented:
I've now renamed the crate to `target-lexicon`, published it to crates.io, and updated the PR here to use it. I've addressed all the comments above. Does anyone have any other concerns?

sunfishcode commented:
Cool. Let's push the big green button :-).
